symengine/sets.cpp

```diff
@@ -418,57 +378,88 @@ RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
     if (is_a<Interval>(*o)) {
         set_basic container;
         for (const auto &a : container_) {
-            if (o->contains(a))
+            auto contain = o->contains(a);
+            if (eq(*contain, *boolTrue))
                 container.insert(a);
+            if (is_a<Contains>(*contain))
+                throw std::runtime_error("Not implemented");
         }
         return finiteset(container);
     }
-    return (*o).set_intersection(rcp_from_this_cast<const Set>());
+    if (is_a<UniversalSet>(*o) || is_a<EmptySet>(*o) || is_a<Union>(*o)) {
+        return (*o).set_intersection(rcp_from_this_cast<const Set>());
+    }
+    throw std::runtime_error("Not implemented");
 }
 
-bool FiniteSet::is_subset(const RCP<const Set> &o) const
+Union::Union(set_basic in) : container_(in)
```

Comment:
It's better to use a `std::set<RCP<const Set>>` because you don't need other types here.